frontend/src/framework/internal/persistence/core/SessionStateTracker.ts

```diff
@@ -0,0 +1,189 @@
+        const backendHash = this._session.getMetadata().hash;
+
+        // If session is persisted and has a backend hash, use it as the source of truth
+        // This applies even if loaded from localStorage (persisted sessions recovered from localStorage)
+        if (this._session.getIsPersisted() && backendHash) {
+            // Use the hash provided by the backend as the source of truth for BOTH current and persisted
+            // This prevents false positives from deserialization artifacts
+            this._state.currentHash = backendHash;
+            this._state.lastPersistedHash = backendHash;
+
+            // Store the current metadata as last persisted since we just loaded from backend
+            this._state.lastPersistedMetadata = {
+                title: this._session.getMetadata().title,
+                description: this._session.getMetadata().description,
+            };
+        } else {
+            // Hash only the content to match backend behavior
+            const contentString = objectToJsonString(this._session.serializeContentState());
+            const hash = await hashSessionContentString(contentString);
+
+            this._state.currentHash = hash;
+
+            // Only set lastPersistedHash/Metadata if the session has been persisted to backend
+            // New unpersisted sessions should have null to indicate they need saving
+            if (this._session.getIsPersisted()) {
+                // Fallback case: persisted session but no hash provided
+                this._state.lastPersistedHash = hash;
+                this._state.lastPersistedMetadata = {
+                    title: this._session.getMetadata().title,
+                    description: this._session.getMetadata().description,
+                };
+            } else {
+                // New unpersisted sessions don't have a trusted backend hash
+                this._state.lastPersistedHash = null;
+                this._state.lastPersistedMetadata = null;
+            }
+        }
+
+        this.updateSnapshot();
+    }
+
+    async refresh(): Promise<boolean> {
+        // Hash only the content to match backend behavior
+        const newContentString = objectToJsonString(this._session.serializeContentState());
+        const newHash = await hashSessionContentString(newContentString);
+
+        // For state change detection, include metadata (title/description) as well
+        const newStateString = makeWorkbenchSessionStateString(this._session);
+
+        if (newHash !== this._state.currentHash) {
+            this._state.currentStateString = newStateString;
+            this._state.currentHash = newHash;
+            this._state.lastModifiedMs = Date.now();
+            this._session.updateMetadata({ lastModifiedMs: this._state.lastModifiedMs }, false);
+            this.updateSnapshot();
+            return true;
+        }
+        return false;
```

Comment:
For readability, switch to early retrurn? `if (newHash === this._state.currentHash) return false;`